### ISSUE TITLE
after master_june24: consider releasing the assumption that two adjacent SIMD vectors have the same channelid in mixed mode

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#926]
See Olivier's comment here

https://github.com/madgraph5/madgraph4gpu/issues/911#issuecomment-2238764093

![image](https://github.com/user-attachments/assets/db562d7f-6b8e-457c-99c3-62c1e09beebe)

**Comment by oliviermattelaer** [comment]
@valassi, indeed we can wait for this, but I would like to understand what the issue is.
(Both to check that I do understand why you want to postpone this, and to be sure that we are on the same page in term of coding)

Is the issue is that you need to change some loop position?

Conceptually, they are no dependence in channelIDs for the color part. (so they are no divergence related to it in the color part)
so if you do (for avx512)
warp_size=16, fptype=m nb_warp=1 this obviously match your assumption and it does work
now you also support
warp_size=16, fptype=m nb_warp=2 and loop twice over the matrix-element in this case and do two SIMD computation for the color (and four SIMD computation for the matrix-element).

for me this should be fully identical compare to 
warp_size=8, fptype=m nb_warp=4
where you also need to do TWO SIMD computation for the color and four SIMD computation for the matrix-element.
Additionally the data structure is the same for both value, so I clearly miss a point.

**Comment by valassi** [comment]
> valassi, indeed we can wait for this, 

Thanks Olivier :-)

>but I would like to understand what the issue is. (Both to check that I do understand why you want to postpone this, and to be sure that we are on the same page in term of coding)

Ok sorry, I thought I had written it down, but it was on a comment I was still writing on the related comment here
https://github.com/madgraph5/madgraph4gpu/issues/898#issuecomment-2230468157

![image](https://github.com/user-attachments/assets/0308de77-58be-40f1-b9af-cd2ff1d1b837)

So I will explain it here instead.

> Is the issue is that you need to change some loop position?
> 
> Conceptually, they are no dependence in channelIDs for the color part. (so they are no divergence related to it in the color part) so if you do (for avx512) warp_size=16, fptype=m nb_warp=1 this obviously match your assumption and it does work now you also support warp_size=16, fptype=m nb_warp=2 and loop twice over the matrix-element in this case and do two SIMD computation for the color (and four SIMD computation for the matrix-element).
> 
> for me this should be fully identical compare to warp_size=8, fptype=m nb_warp=4 where you also need to do TWO SIMD computation for the color and four SIMD computation for the matrix-element. Additionally the data structure is the same for both value, so I clearly miss a point.

There is indeed no a-priori reason why in mixed mode two SIMD vectors should have the same channelId. Example avx2, 4 doubles per vector Feynman, 8 floats per vector in color, you may choose to allow channelids being the same over 8 elements or over 4. Both can be implemented. But **these are two different choices, which require different implementation code and different SANITY CHECKs**. The implementation in 830 was tatally missing any internal consistency and sanity checks on these points. Specifically the two choices are
- One, what I implemented now, because it is the closest to upstream/master and reuses most of its code. You require two 4-vectors (8 elements) to have the same channel. The implementation has calculate_wavefunction *internally* iterate over two vectors, and it takes a single scalar channelId value. Then in sigmakin I just add the sanity check that indeed 8 elements have the same channelid, before passing them to calculate_wavefunctions.
- Two, one can allow the two 4-vectors to have two different channelIds. The easiest here wuld be to pass TWO channelId scalars to calculate_wavefunction in mixed mode, and let this proceed accordingly.

(About why it is better to have calculate_wavefunction take a scalar channelid and not an array, I will comment in #894. The SIMD implementation inside calculate_wavefunction, if a channelid array is allowed, was completelywrong. Again, any design choice is possible. But then all pieces of the code must be internally consistent, and check any hidden assumptions.)

On the implementation two, what you are suggesting, it would not take a huge amount of work, but it would take some. Essentially this was the working assumption I was using yesterday. Have a look at this reverse-commit to see some parts of what it would take
https://github.com/valassi/madgraph4gpu/commit/e7f38529bc83aebb179a714e2d7cc428caaf868e
```
git diff --no-ext-diff e7f38529bc83aebb179a714e2d7cc428caaf868e..e7f38529bc83aebb179a714e2d7cc428caaf868e~
diff --git a/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx/CPPProcess.cc b/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx/CPPProcess.cc
index df51b3c61..7b503e35e 100644
--- a/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx/CPPProcess.cc
+++ b/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx/CPPProcess.cc
@@ -1083,10 +1083,12 @@ namespace mg5amcCpu
       const int ievt00 = ipagV2 * neppV; // loop on one SIMD page (neppV events) at a time
 #endif
 #ifdef MGONGPU_SUPPORTS_MULTICHANNEL
-      // SCALAR channelId for the whole SIMD neppV2 event page (C++), i.e. one or two neppV event page(s)
-      // The cudacpp implementation ASSUMES (and checks! #898) that all channelIds are the same in a neppV2 SIMD event page
-      // **NB! in "mixed" precision, using SIMD, calculate_wavefunctions computes MEs for TWO neppV pages with a single channelId! #924
+      // SCALAR channelId for the current event (CUDA) or for the whole SIMD neppV event page(s) (C++)
+      // The cudacpp implementation ASSUMES (and checks! #898) that all channelIds are the same in a neppV SIMD event page
       unsigned int channelId = 0; // disable multichannel single-diagram enhancement unless allChannelIds != nullptr
+#if defined MGONGPU_CPPSIMD and defined MGONGPU_FPTYPE_DOUBLE and defined MGONGPU_FPTYPE2_FLOAT
+      unsigned int channelId2 = 0; // disable multichannel single-diagram enhancement unless allChannelIds != nullptr
+#endif
       if( allChannelIds != nullptr )
       {
         // First - and/or only - neppV page of channels (iParity=0 => ievt0 = ievt00 + 0 * neppV)
@@ -1107,12 +1109,14 @@ namespace mg5amcCpu
 #if defined MGONGPU_CPPSIMD and defined MGONGPU_FPTYPE_DOUBLE and defined MGONGPU_FPTYPE2_FLOAT
         // Second neppV page of channels (iParity=1 => ievt0 = ievt00 + 1 * neppV)
         const unsigned int* channelIds2 = CID_ACCESS::ieventAccessRecordConst( allChannelIds, ievt00 + neppV ); // fix bug #899/#911
-        uint_v channelIds2_v = CID_ACCESS::kernelAccessConst( channelIds2 );                                    // fix #895 (compute this only once for all diagrams)
-        // **NB! in "mixed" precision, using SIMD, calculate_wavefunctions computes MEs for TWO neppV pages with a single channelId! #924
-        for( int i = 0; i < neppV; ++i )
+        uint_sv channelIds2_sv = CID_ACCESS::kernelAccessConst( channelIds2 );                                  // fix #895 (compute this only once for all diagrams)
+        // NB: channelIds_sv is a vector in SIMD C++
+        channelId2 = channelIds2_sv[0];  // element[0]
+        for( int i = 1; i < neppV; ++i ) // elements[1...neppV-1]
         {
-          assert( channelId == channelIds2_v[i] ); // SANITY CHECKS #898 #924: all events in the 2nd SIMD vector have the same channelId as that of the 1st SIMD vector
+          assert( channelId2 == channelIds2_sv[i] ); // SANITY CHECK #898: check that all events in a SIMD vector have the same channelId
         }
+        assert( channelId2 > 0 ); // SANITY CHECK: scalar channelId must be > 0 if multichannel is enabled (allChannelIds != nullptr)
 #endif
       }
 #endif
@@ -1127,7 +1131,6 @@ namespace mg5amcCpu
       {
         const int ihel = cGoodHel[ighel];
 #ifdef MGONGPU_SUPPORTS_MULTICHANNEL
-        // **NB! in "mixed" precision, using SIMD, calculate_wavefunctions computes MEs for TWO neppV pages with a single channelId! #924
         calculate_wavefunctions( ihel, allmomenta, allcouplings, allMEs, channelId, allNumerators, allDenominators, jamp2_sv, ievt00 );
 #else
         calculate_wavefunctions( ihel, allmomenta, allcouplings, allMEs, jamp2_sv, ievt00 );
```

In practice
- you would need to compute two scalars, channelId and channelId2
- what is missing in the snippet here is that for FPTYPE=m one ifdef should add asecond channelId parameter to calculate_wavefunction

Not a massive amount of work, but I'd rather not do that now.

That said, on the use case, honestly I am not sure this will make such a difference. **And this is why I did not implement this**. Ok it allows us to send half the number of events in a job if we really want to squeeze things. But **will we ever ONLY calculate 4 events in a given channel for a job?** I really doubt it. If you consider that for a given channel one madeven executable will in any case process maybe 100 events, whether you do that with warp_size 4 or warp_size 8 or even 16 or 32 frankly I think makes no difference.

I hope this explains.

**Comment by oliviermattelaer** [comment]
Ok thanks,

So long story short, this boils down to the fact that channelIDs is used as a scalar for the SIMD case which is ok (as long as this is not the case for GPU which seems what you refer to in #898).

**Comment by valassi** [comment]
> Ok thanks,
> 
> So long story short, this boils down to the fact that channelIDs is used as a scalar for the SIMD case which is ok (as long as this is not the case for GPU which seems what you refer to in #898).

Yes it is used as a scalar in SIMD, and in mixed precision SIMD you would need two scalars.

Note, GPU is a different story. **I did not add any sanity checks for GPU that the channelId in a warp are the same** (please tell me if you think this should be done). I should cross check again, but IIRC sigmakin is a kernel and runs once per thread in a warp, and identifies a different channelid in each thread anyway, which is then passed to calculate_wavefunction. THis means that
- **functionally**, it will work and give the right results even if the channelids in a warp are not the same
- **performance wise**, of course it is better to avoid divergence and always use reasonable WARP_SIZE which matches the hardware
- but strictly speaking sanity checks are not needed for the functionality... if you want we can add them to ensure performance however, let me know (it can be done later in two lines of code)
